aip/client-libraries/4234.md

```diff
@@ -0,0 +1,106 @@
+product service that hosts the mix-in service.
+
+If client libraries support mix-in services, they **must** support the following
+common services (and **may** support others):
+
+- `google.cloud.location.Locations`
+- `google.iam.v1.IAMPolicy`
+- `google.longrunning.Operations`
+
+**Note:** The list of supported mix-in services and the mix-in services
+themselves change infrequently. Any new common service to be supported should be
+added to the list here and generators should be updated.
+
+To be generated into a client library, a mix-in service **must** be declared
+under [apis] in a [google.api.Service]. Furthermore, it **must** have
+[google.api.http] bindings declared for all of its RPCs in the [http]
+configuration of the same [google.api.Service] (more details on these bindings
+in AIP-127).
+
+
+## Implementing mix-in support
+
+### Generator configuration
+
+Client library generators **must** accept the file path of a
+[google.api.Service] in YAML form. This flag **must** be optional - meaning
+the flag **must not** be required for basic client generation that omits the
+mix-in methods.
+
+### Mix-in API client configuration
+
+If a mix-in API utilizes client library configuration such as
+`google.api.method_signature` or request header injection, generators **may**
+support it in the generated mix-in methods if the configuration is accessible.
+
+### Generating mix-in methods
+
+The mix-in API RPCs **should** be generated as methods on the surface of the
+host API's service client library, alongside the host service's RPCs. This
+presents them as top-level methods on the client, but under-the-hood, the
+appropriate mix-in API stub or request should be invoked.
+
+**Note:** For gRPC clients, the mix-in API gRPC stub **must** be used in order
+to properly construct the gRPC request.
+
+### Multiple host service clients
+
+If there are multiple services defined by the host API that would result in
+multiple clients being generated, generators **must** include the mix-in
+methods on all eligible, generated service clients.
+
+### Overriding a duplicate RPC
+
+Client library generators **must not** generate a mix-in method on a host
+service client if that service already defines an RPC with the same name. For
+example, take the following service definition:
+
+```proto
+service LibraryService {
+  rpc ListBooks(ListBookRequest) returns (ListBooksResponse);
+
+  // Other host API RPCs...
+
+  // Redefinitions of the google.iam.v1.IAMPolicy mix-in service.
+  rpc GetIamPolicy(google.iam.v1.GetIamPolicyRequest)
+      returns (google.iam.v1.Policy);
+
+  rpc SetIamPolicy(google.iam.v1.SetIamPolicyRequest)
+      returns (google.iam.v1.Policy);
+
+  rpc TestIamPermissions(google.iam.v1.TestIamPermissionsRequest)
+      returns (google.iam.v1.TestIamPermissionsResponse);
+}
+```
+
+Should the host API declare the `google.iam.v1.IAMPolicy` as a mix-in service,
+client library generators **must not** override the methods generated for the
```

Comment:
Consider adding handling guidance for APIs who may have two services, one of which overrides an RPC. Example use case:
- PubSub mixes in IAM, and has Publisher and Subscriber services.
- Subscriber defines its own `GetIamPolicy`.
- What does Publisher do?

Proposal: IAM's `GetIamPolicy` should not be mixed into the Publisher client, which is based on how existing APIs copy-paste IAM protos into their own.

Reply:
I'm not super opinionated here.

A reason I am _for_ your proposal is that X service has the desired RPC overrides, and any representation of them on another service via mixins might run counter to some customization the overrides have. I guess it just requires checking all methods to be generated and if any service has the overrides, ignore IAMPolixy as a mixin.

Reply:
@miraleung why do you think Publisher shouldn't have the mixin version of the RPCs?

Reply:
Rationale: If PubSub is overriding `GetIamPolicy`, then they are responsible for defining whether its full surface. Perhaps there's a reason for defining it only in Subscriber but not Publisher? Besides, if we automatically mix `GetIamPolicy` into Publisher, how would the team turn that off?

Given that this use case doesn't actually exist yet, I don't have a strong preference. IMO, the only remaining key factor is that it is easier to add the overridden RPCs back in if this proposal doesn't work out, than to go the opposite direction.

Reply:
That's fair! 

> it is easier to add the overridden RPCs back in if this proposal doesn't work out,

Yes, great point. I'm sold. I will update this section as you've suggested.

Reply:
Done.